gpflow/models/sgpr.py

```diff
@@ -150,46 +151,95 @@ class SGPR(SGPRBase):
+
+        return self.CommonTensors(A, B, LB, AAT, L)
+    
+    def _logdet_term(self, common):
+        """
+        Bound from Jensen's Inequality:
+
+        log |K + σ²I| <= log |Q + σ²I| + N * log (1 + tr(K - Q)/(σ²N))
+
+        :param B: Matrix returned from common calc, unused
+        :param LB: Matrix returned from common calc
+        :param AAT: Matrix returned from common calc
+        :return: log_det, upper bound on .5 * log |K + σ²I|
+        """
+        LB = common.LB
+        AAT = common.AAT
+
+        x, y = self.data
+        num_data = to_default_float(tf.shape(x)[0])
+        output_dim = to_default_float(tf.shape(y)[1])
+        kdiag = self.kernel(x, full_cov=False)
+        sigma_sq = self.likelihood.variance
+
+        # tr(K - Q) / σ²
+        trace = tf.reduce_sum(kdiag) / sigma_sq - tf.reduce_sum(tf.linalg.diag_part(AAT))
+
+        # log(det(B))
+        logdet_b = output_dim * tf.reduce_sum(tf.math.log(tf.linalg.diag_part(LB)))
+
+        # N * log(\sigma^2)
+        log_sigma_sq = num_data * output_dim * tf.math.log(sigma_sq)
+
+        logdet_k = - (logdet_b + 0.5 * log_sigma_sq + 0.5 * trace)
+        return logdet_k
+    
+    def _quad_term(self, common) -> tf.Tensor:
+        A = common.A
+        LB = common.LB
+
+        x, y = self.data
+        err = y - self.mean_function(x)
+        sigma_sq = self.likelihood.variance
+        sigma = tf.sqrt(sigma_sq)
+
         Aerr = tf.linalg.matmul(A, err)
         c = tf.linalg.triangular_solve(LB, Aerr, lower=True) / sigma
 
-        # compute log marginal bound
-        bound = -0.5 * num_data * output_dim * np.log(2 * np.pi)
-        bound += tf.negative(output_dim) * tf.reduce_sum(tf.math.log(tf.linalg.diag_part(LB)))
-        bound -= 0.5 * num_data * output_dim * tf.math.log(self.likelihood.variance)
-        bound += -0.5 * tf.reduce_sum(tf.square(err)) / self.likelihood.variance
-        bound += 0.5 * tf.reduce_sum(tf.square(c))
-        bound += -0.5 * output_dim * tf.reduce_sum(Kdiag) / self.likelihood.variance
-        bound += 0.5 * output_dim * tf.reduce_sum(tf.linalg.diag_part(AAT))
+        quad = -0.5 * (tf.reduce_sum(tf.square(err)) / sigma_sq - tf.reduce_sum(tf.square(c)))
+        return quad
 
-        return bound
+    def elbo(self) -> tf.Tensor:
+        """
+        Construct a tensorflow function to compute the bound on the marginal
+        likelihood. For a derivation of the terms in here, see the associated
+        SGPR notebook.
+        """
+        common = self._common_calculation()
+        num_data = to_default_float(tf.shape(self.data[-1])[0])
+        output_dim = to_default_float(tf.shape(self.data[-1])[1])
+        const = -0.5 * num_data * output_dim * np.log(2 * np.pi)
+        logdet = self._logdet_term(common)
+        quad = self._quad_term(common)
+        return const + logdet + quad
 
     def predict_f(self, Xnew: InputData, full_cov=False, full_output_cov=False) -> MeanAndVariance:
```

Comment:
`predict_f` could also make use of `_common_calculation` to remove redundant code I think.

Reply:
I would prefer to minimise changes in SGPR, and narrow the scope of this PR to CGLB only.